llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp

```diff
@@ -2694,19 +2694,49 @@ bool SPIRVInstructionSelector::selectIntrinsic(Register ResVReg,
     }
     return MIB.constrainAllUses(TII, TRI, RBI);
   }
-  case Intrinsic::spv_loop_merge:
-  case Intrinsic::spv_selection_merge: {
-    const auto Opcode = IID == Intrinsic::spv_selection_merge
-                            ? SPIRV::OpSelectionMerge
-                            : SPIRV::OpLoopMerge;
-    auto MIB = BuildMI(BB, I, I.getDebugLoc(), TII.get(Opcode));
+  case Intrinsic::spv_loop_merge: {
+    auto MIB = BuildMI(BB, I, I.getDebugLoc(), TII.get(SPIRV::OpLoopMerge));
     for (unsigned i = 1; i < I.getNumExplicitOperands(); ++i) {
       assert(I.getOperand(i).isMBB());
       MIB.addMBB(I.getOperand(i).getMBB());
     }
     MIB.addImm(SPIRV::SelectionControl::None);
     return MIB.constrainAllUses(TII, TRI, RBI);
   }
+  case Intrinsic::spv_selection_merge: {
+
+    auto SelectionControl = SPIRV::SelectionControl::None;
+    auto LastOp = I.getOperand(I.getNumExplicitOperands() - 1);
```

Comment:
This code still makes an assumption that if metadata is present on this instruction it must be HLSL branch metadata... That seems like an unsafe assumption to make. Shouldn't we force that we're reading the `hlsl.controlflow.hint` metadata.

Reply:
Stepping back. First off, we don't actually have a test showing the IR post-SPIRVStructurizer, so that part of your change isn't unit tested. The transformation performed in that pass actually might be the root of my concern here.

It looks like you're translating the metadata from the terminator instruction into an argument to a SPIR-V intrinsic, then lowering that intrinsic. That isn't inherently bad, but it makes it a really strange decision to keep the metadata in its language-specific named metadata form as an argument to an intrinsic.

It seems like we should probably translate the metadata into an immediate value passed to the intrinsic rather than keeping it as metadata.

Reply:
@Keenuts @s-perron, I am trying to implement this suggestion from @llvm-beanz, but I am facing an issue.

This is the current definition for `OpSelectMerge`:
```C++
  def int_spv_loop_merge : Intrinsic<[], [llvm_vararg_ty]>;
```
The suggested solution, requires me to change into this:
 
```C++
  def int_spv_selection_merge : Intrinsic<[], [llvm_i32_ty,llvm_vararg_ty], [ImmArg<ArgIndex<0>>]>;
```
The problem I am facing is, the current first parameter of `OpSelectMerge` is a basic block address, and there is a bunch of places that do similar to the following:
 
```C++
  // Otherwise, we need to check: was there an OpSelectionMerge before this
  // branch? If we removed the OpBranchConditional, we must also remove the
  // OpSelectionMerge. This is not valid for OpLoopMerge:
  IntrinsicInst *II =
      dyn_cast<IntrinsicInst>(BB->getTerminator()->getPrevNode());
  if (!II || II->getIntrinsicID() != Intrinsic::spv_selection_merge)
    return;
  Constant *C = cast<Constant>(II->getOperand(0));
  II->eraseFromParent();
  if (!C->isConstantUsed())
    C->destroyConstant();
}
```

My question is, should I go through the code and replace everywhere that is reading the first operand to read the second? Or there is a better way to pass this value to op instruct selector?

Reply:
> ```c++
>   def int_spv_loop_merge : Intrinsic<[], [llvm_vararg_ty]>;
> ```

Assuming you meant to mention only `OpSelectionMerge` and not `OpLoopMerge`.


> The suggested solution, requires me to change into this:
> ```c++
>   def int_spv_selection_merge : Intrinsic<[], [llvm_i32_ty,llvm_vararg_ty], [ImmArg<ArgIndex<0>>]>;
> ```

`OpSelectionMerge` has a fixed number of operands [(spec)](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpSelectionMerge):
This means you should be able to do this:

```
def int_spv_selection_merge : Intrinsic<[], [llvm_any_ty, llvm_i32_ty]>
```

Respecting the order of operands of the SPIR-V instruction, hence you shouldn't have to modify the code dealing with it.
So it looks like if you change the intrinsic to have those fixed arguments, you should be able to emit the mask in the frontend when generating the intrinsic call. If you do so, this lowering code will become way simpler.

Reply:
@llvm-beanz @Keenuts SPIRV Structurizer pass is not register in `opt` CLI. Also, I don't see any tests, only calling the structurizer. That being said, there is currently [this test](https://github.com/llvm/llvm-project/pull/116331/files#diff-b48f32fed30eadb1a750bf5cd542c92d33824c63df80bbc86eafc16b175a316e), which follow the same pattern as other structurizer tests. 

Is such test enough? If any more tests are needed, can you give me an example on how to do it, please?

Reply:
Made a separate PR to add the structurizer into opt https://github.com/llvm/llvm-project/pull/119301/files